src/utils/parse.rs

```diff
@@ -83,3 +84,286 @@ where
     }
     Ok(vector)
 }
+
+#[derive(PartialEq, Debug)]
+enum ParseState {
+    Start,
+    LeftBracket,
+    Index,
+    Value,
+    Splitter,
+    Comma,
+    Length,
+}
+
+#[inline(always)]
+pub fn svector_filter_nonzero<T: Zero + Clone + PartialEq>(
+    indexes: &[u32],
+    values: &[T],
+) -> (Vec<u32>, Vec<T>) {
+    let non_zero_indexes: Vec<u32> = indexes
+        .iter()
+        .enumerate()
+        .filter(|(i, _)| values.get(*i).unwrap() != &T::zero())
+        .map(|(_, x)| *x)
+        .collect();
+    let non_zero_values: Vec<T> = indexes
+        .iter()
+        .enumerate()
+        .filter(|(i, _)| values.get(*i).unwrap() != &T::zero())
+        .map(|(i, _)| values.get(i).unwrap().clone())
+        .collect();
+    (non_zero_indexes, non_zero_values)
+}
+
+#[inline(always)]
+pub fn parse_pgvector_svector<T: Zero + Clone, F>(
+    input: &[u8],
+    f: F,
+) -> Result<(Vec<u32>, Vec<T>, usize), ParseVectorError>
+where
+    F: Fn(&str) -> Option<T>,
+{
+    use arrayvec::ArrayVec;
+    if input.is_empty() {
+        return Err(ParseVectorError::EmptyString {});
+    }
+    let mut token: ArrayVec<u8, 48> = ArrayVec::new();
+    let mut indexes = Vec::<u32>::new();
+    let mut values = Vec::<T>::new();
+
+    let mut state = ParseState::Start;
+    for (position, char) in input.iter().enumerate() {
+        let c = *char;
+        match (&state, c) {
+            (_, b' ') => {}
+            (ParseState::Start, b'{') => {
+                state = ParseState::LeftBracket;
+            }
+            (
+                ParseState::LeftBracket | ParseState::Index | ParseState::Comma,
+                b'0'..=b'9' | b'a'..=b'z' | b'A'..=b'Z' | b'.' | b'+' | b'-',
+            ) => {
+                if token.is_empty() {
+                    token.push(b'$');
+                }
+                if token.try_push(c).is_err() {
+                    return Err(ParseVectorError::TooLongNumber { position });
+                }
+                state = ParseState::Index;
+            }
+            (ParseState::LeftBracket | ParseState::Comma, b'}') => {
+                state = ParseState::Splitter;
+            }
+            (ParseState::Index, b':') => {
+                if token.is_empty() {
+                    return Err(ParseVectorError::TooShortNumber { position });
+                }
+                let s = unsafe { std::str::from_utf8_unchecked(&token[1..]) };
+                let index = s
+                    .parse::<u32>()
+                    .map_err(|_| ParseVectorError::BadParsing { position })?;
+                indexes.push(index);
+                token.clear();
+                state = ParseState::Value;
```

Comment:
There should be a state, if the last input character is `:`. And we do not need to check `token.is_empty()` because if parser does not see digitals it does not enter `Index`, `Value` or `Dims`.